crates/extensions/c8y_mapper_ext/src/operations/mod.rs

```diff
@@ -13,47 +13,270 @@
+                    }
+                    OperationType::ConfigUpdate => {
+                        handler
+                            .handle_config_update_state_change(entity, &cmd_id, &message)
+                            .await
+                    }
+                    OperationType::FirmwareUpdate => {
+                        handler
+                            .handle_firmware_update_state_change(entity, &cmd_id, &message)
+                            .await
+                    }
+                };
+
+                let mut mqtt_publisher = handler.mqtt_publisher.clone();
+                match res {
+                    // If there are mapped final status messages to be published, they are cached until the operation
+                    // log is uploaded
+                    Ok((messages, command)) => {
+                        if let Some(command) = command {
+                            if let Err(e) = handler
+                                .upload_operation_log(&external_id, &cmd_id, &operation, command)
+                                .await
+                            {
+                                error!("failed to upload operation logs: {e}");
+                            }
+                        }
+
+                        for message in messages {
+                            // if task publishes MQTT clearing message, an operation is considered finished, so we can
+                            // terminate the task as well
+                            if message.retain
+                                && message.payload_bytes().is_empty()
+                                && message.topic == command_topic
+                            {
+                                rx.close();
+                            }
+                            mqtt_publisher.send(message).await.unwrap();
+                        }
+                    }
+                    Err(e) => error!("{e}"),
+                }
+            }
+        });
 
-    pub entity_topic_id: EntityTopicId,
+        Self { handle, tx }
+    }
+}
 
-    pub command: GenericCommandState,
+/// An MQTT message that contains an operation payload.
+///
+/// These are MQTT messages that contain operation payloads. These messages need to be passed to
+/// tasks that handle a given operation to advance the operation and eventually complete it.
+struct OperationMessage {
+    operation: OperationType,
+    entity: Entity,
+    cmd_id: Arc<str>,
+    message: MqttMessage,
 }
 
-/// Used to denote as type of what operation was the file downloaded from the FTS.
+/// A subset of entity-related information necessary to handle an operation.
 ///
-/// Used to dispatch download result to the correct operation handler.
-pub enum FtsDownloadOperationType {
-    LogDownload,
-    ConfigDownload,
+/// Because the operation may take time and other operations may run concurrently, we don't want to
+/// query the entity store.
+#[derive(Clone, Debug)]
+pub struct Entity {
```

Comment:
```suggestion
pub struct EntityTarget {
```
or `EntityInfo` or anything similar to avoid any confusion with the general concept of entity.